src/client/Microsoft.Identity.Client/ManagedIdentity/V2/WindowsPersistentCertificateCache.cs

```diff
@@ -0,0 +1,336 @@
+                        }
+
+                        if (candidate.NotAfter > bestNotAfter)
+                        {
+                            best?.Dispose();
+                            best = new X509Certificate2(candidate); // caller-owned clone (preserves private key link)
+                            bestEndpoint = endpointBase;
+                            bestNotAfter = candidate.NotAfter;
+                        }
+                    }
+                    finally
+                    {
+                        candidate.Dispose();
+                    }
+                }
+
+                if (best != null)
+                {
+                    // CN (GUID) → canonical client_id
+                    string cn = null;
+                    try
+                    {
+                        cn = best.GetNameInfo(X509NameType.SimpleName, false);
+                    }
+                    catch (Exception ex)
+                    {
+                        logger?.Verbose(() => "[PersistentCert] Failed to read CN from selected certificate: " + ex.Message);
+                    }
+
+                    if (!Guid.TryParse(cn, out var clientIdGuid))
+                    {
+                        best.Dispose();
+                        logger?.Verbose(() => "[PersistentCert] Selected entry CN is not a GUID; skipping.");
+                        return false;
+                    }
+
+                    value = new CertificateCacheValue(best, bestEndpoint, clientIdGuid.ToString("D"));
+                    logger?.Verbose(() => "[PersistentCert] Reused certificate from CurrentUser/My.");
+                    return true;
+                }
+            }
+            catch (Exception ex)
+            {
+                logger?.Verbose(() => "[PersistentCert] Store lookup failed: " + ex.Message);
+            }
+
+            return false;
+        }
+
+        public void Write(string alias, X509Certificate2 cert, string endpointBase, ILoggerAdapter logger = null)
+        {
+            if (!PlatformSupportsPersistentFriendlyName(logger) || cert == null)
+                return;
+
+            // IMDSv2 attaches the private key earlier (will throw if it cannot).
+            // We do not block here; log defensively if we see a public-only cert.
+            if (!cert.HasPrivateKey)
+            {
+                logger?.Verbose(() => "[PersistentCert] Unexpected: Write() received a cert without a private key. Continuing best-effort.");
+            }
+
+            if (!MsiCertificateFriendlyNameEncoder.TryEncode(alias, endpointBase, out var friendlyName))
+            {
+                logger?.Verbose(() => "[PersistentCert] FriendlyName encode failed; skipping persist.");
+                return;
+            }
+
+            // Best-effort: short lock, skip if busy
+            InterprocessLock.TryWithAliasLock(
+                alias,
+                timeout: TimeSpan.FromMilliseconds(300),
```

Comment:
Should this be configurable? Should we add a retry loop?

Reply:
Kept a single short, non‑configurable timeout and no retry, and added a comment. Persistent store access is explicitly best‑effort and must not block token acquisition; if the lock is busy we skip the write/delete and fall back to the in‑memory cache instead of retrying.